catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/MetacardTypeImpl.java

```diff
@@ -157,13 +160,14 @@ public AttributeDescriptor getAttributeDescriptor(String attributeName) {
     if (attributeName == null) {
       return null;
     }
-    // TODO could this be faster?
-    for (AttributeDescriptor descriptor : descriptors) {
-      if (attributeName.equals(descriptor.getName())) {
-        return descriptor;
+
+    if (descriptors.size() != descriptorNameMap.size()) {
```

Comment:
If the above is the case:
✏️ perhaps you could use a flag variable like `shouldUpdateNameMap`, set that to true every time the set of attribute descriptors is modified, then check the flag here, update the name map and reset the flag if needed. 
That way you wouldn't have to check through to see if all of the attributes are actually the same.

Alternatively, you might be able to add this logic to the end of `addAttributeDescriptors` and refactor the one constructor to use `addAttributeDescriptors` instead of `addAll()`.

Reply:
@coyotesqrl The case could occur, but I don't think it's too likely. Most of the time subclasses are just adding descriptors. 
@zta6 I don't know of any to add a listener to a hashset that fires when the set changes. The problem is subclasses are directly adding descriptors via `descriptors.addAll()`

Reply:
My opinion would be that it's better to be slow and always correct